pkg/keyspace/keyspace.go

```diff
@@ -325,65 +328,62 @@ func (manager *Manager) splitKeyspaceRegion(id uint32, waitRegionSplit bool) (er
 	defer func() {
 		if err != nil {
 			cl.GetRegionLabeler().DeleteLabelRule(keyspaceRule.ID)
+			return
 		}
+		log.Info("[keyspace] added region label for keyspace",
+			zap.Uint32("keyspace-id", id),
+			zap.Any("label-rule", keyspaceRule),
+			zap.Duration("takes", time.Since(start)),
+		)
 	}()
 
-	if waitRegionSplit {
-		ranges := keyspaceRule.Data.([]*labeler.KeyRangeRule)
-		if len(ranges) < 2 {
-			log.Warn("[keyspace] failed to split keyspace region with insufficient range", zap.Any("label-rule", keyspaceRule))
-			return ErrRegionSplitFailed
-		}
-		rawLeftBound, rawRightBound := ranges[0].StartKey, ranges[0].EndKey
-		txnLeftBound, txnRightBound := ranges[1].StartKey, ranges[1].EndKey
+	if !waitRegionSplit {
+		return
+	}
 
-		ticker := time.NewTicker(manager.config.GetCheckRegionSplitInterval())
-		timer := time.NewTimer(manager.config.GetWaitRegionSplitTimeout())
-		defer func() {
-			ticker.Stop()
-			timer.Stop()
-		}()
-		for {
-			select {
-			case <-ticker.C:
-				regionsInfo := manager.cluster.GetBasicCluster().RegionsInfo
-				region := regionsInfo.GetRegionByKey(rawLeftBound)
-				if region == nil || !bytes.Equal(region.GetStartKey(), rawLeftBound) {
-					continue
-				}
-				region = regionsInfo.GetRegionByKey(rawRightBound)
-				if region == nil || !bytes.Equal(region.GetStartKey(), rawRightBound) {
-					continue
-				}
-				region = regionsInfo.GetRegionByKey(txnLeftBound)
-				if region == nil || !bytes.Equal(region.GetStartKey(), txnLeftBound) {
-					continue
-				}
-				region = regionsInfo.GetRegionByKey(txnRightBound)
-				if region == nil || !bytes.Equal(region.GetStartKey(), txnRightBound) {
-					continue
+	// Stat waiting for region split.
+	var splitKeys [][]byte
+	for _, regionRange := range keyspaceRule.Data.([]*labeler.KeyRangeRule) {
+		splitKeys = append(splitKeys, regionRange.StartKey, regionRange.EndKey)
+	}
+	tickerInterval := manager.config.GetCheckRegionSplitInterval()
+	ticker := time.NewTicker(tickerInterval)
+	timer := time.NewTimer(manager.config.GetWaitRegionSplitTimeout())
+	defer func() {
+		ticker.Stop()
+		timer.Stop()
+	}()
+	for {
+		select {
+		case <-ticker.C:
+			// Reset ticker if it's interval has been changed.
+			newInterval := manager.config.GetCheckRegionSplitInterval()
+			if tickerInterval != newInterval {
+				tickerInterval = newInterval
+				ticker.Reset(newInterval)
+			}
+			regionsInfo := manager.cluster.GetBasicCluster().RegionsInfo
+			totalSplit := 0 // number of split keys that have been split.
+			var region *core.RegionInfo
+			for _, splitKey := range splitKeys {
+				region = regionsInfo.GetRegionByKey(splitKey)
+				if region != nil && bytes.Equal(region.GetStartKey(), splitKey) {
```

Comment:
Do we need to check if the region and key being compared each time are duplicated?

Reply:
I don't think that's necessary, the `totalSplit` reset to 0 every tick anyway, so just counting the number should work.

Reply:
If some Regions never split for some reason, will this check last forever?

Reply:
there is a timer controlling the timeout, if it take too long to split, it will throw ErrRegionSplitTimeout
https://github.com/tikv/pd/blob/a7dd9fa89c8c72da1f6bbb096a356f23694de390/pkg/keyspace/keyspace.go#L378